src/Form/Tag.elm

```diff
@@ -0,0 +1,188 @@
+module Form.Tag exposing
+    ( FormType(..)
+    , Model
+    , Msg
+    , ParentMsg(..)
+    , initAddFormModel
+    , initEditFormModel
+    , update
+    , view
+    )
+
+import Data.Tag as Tag
+import Dict exposing (Dict)
+import Html exposing (Html)
+import Html.Attributes as Attributes
+import Ui.Button as Button
+import Ui.Form as Form
+import Ui.Input as Input
+import Util.Validate as Validate
+
+
+type Model
+    = Model ModelData
+
+
+type alias ModelData =
+    { errors : Errors
+    , tag : Tag.Model
+    , originalTag : Maybe Tag.Model
+    }
+
+
+type alias Errors =
+    Dict String Validate.Errors
+
+
+type FormType
+    = Add
+    | Edit
+
+
+type InputId
+    = Name
+    | Value
+
+
+inputIdToString : InputId -> String
+inputIdToString inputId =
+    case inputId of
+        Name ->
+            "name"
+
+        Value ->
+            "value"
+
+
+emptyErrors : Errors
+emptyErrors =
+    Dict.empty
+
+
+initAddFormModel : Model
+initAddFormModel =
+    Model
+        { errors = emptyErrors
+        , tag = Tag.emptyTag
+        , originalTag = Nothing
+        }
+
+
+initEditFormModel : Tag.Model -> Model
+initEditFormModel tag =
+    Model
+        { errors = emptyErrors
+        , tag = tag
+        , originalTag = Just tag
+        }
+
+
+type Msg
+    = InsertedInputValue InputId String
+    | ClickedSubmit
+
+
+type ParentMsg
+    = NoUpdate
+    | SubmittedForm Tag.Model
+
+
+update : Tag.Tags -> Msg -> Model -> ( Model, Cmd Msg, ParentMsg )
```

Comment:
Interesting approach for making this able to communicate upwards. Do you know of other solutions? What will be their pros and cons?